gobblin-utility/src/main/java/gobblin/util/PullFileLoader.java

```diff
@@ -0,0 +1,257 @@
+      this.extensions = Lists.newArrayList();
+      for (String ext : extensions) {
+        this.extensions.add(ext.startsWith(".") ? ext : "." + ext);
+      }
+    }
+
+    @Override
+    public boolean accept(final Path path) {
+      Predicate<String> predicate = new Predicate<String>() {
+        @Override
+        public boolean apply(@Nullable String input) {
+          return path.getName().toLowerCase().endsWith(input);
+        }
+      };
+      return Iterables.any(this.extensions, predicate);
+    }
+  }
+
+  public PullFileLoader(Path rootDirectory, FileSystem fs, Collection<String> javaPropsPullFileExtensions,
+      Collection<String> hoconPullFileExtensions) {
+
+    Set<String> commonExtensions = Sets.intersection(Sets.newHashSet(javaPropsPullFileExtensions),
+        Sets.newHashSet(hoconPullFileExtensions));
+    Preconditions.checkArgument(commonExtensions.isEmpty(),
+        "Java props and HOCON pull file extensions intersect: " + Arrays.toString(commonExtensions.toArray()));
+
+    this.rootDirectory = rootDirectory;
+    this.fs = fs;
+    this.javaPropsPullFileFilter = new ExtensionFilter(javaPropsPullFileExtensions);
+    this.hoconPullFileFilter = new ExtensionFilter(hoconPullFileExtensions);
+  }
+
+  /**
+   * Load a single pull file.
+   * @param path The {@link Path} to the pull file to load.
+   * @param sysProps A {@link Config} used as fallback.
+   * @param loadGlobalProperties if true, will also load at most one *.properties file per directory from the
+   *          {@link #rootDirectory} to the pull file {@link Path}.
+   * @return The loaded {@link Config}.
+   * @throws IOException
+   */
+  public Config loadPullFile(Path path, Config sysProps, boolean loadGlobalProperties) throws IOException {
+    Config fallback = loadGlobalProperties ? loadAncestorGlobalConfigs(path, sysProps) : sysProps;
+
+    if (this.javaPropsPullFileFilter.accept(path)) {
+      return loadJavaPropsWithFallback(path, fallback).resolve();
+    } else if (this.hoconPullFileFilter.accept(path)) {
+      return loadHoconConfigAtPath(path).withFallback(fallback).resolve();
+    } else {
+      throw new IOException(String.format("Cannot load pull file %s due to unrecognized extension.", path));
+    }
+  }
+
+  /**
+   * Find and load all pull files under a base {@link Path} recursively.
+   * @param path base {@link Path} where pull files should be found recursively.
+   * @param sysProps A {@link Config} used as fallback.
+   * @param loadGlobalProperties if true, will also load at most one *.properties file per directory from the
+   *          {@link #rootDirectory} to the pull file {@link Path} for each pull file.
+   * @return The loaded {@link Config}s.
+   */
+  public Collection<Config> loadPullFilesRecursively(Path path, Config sysProps, boolean loadGlobalProperties) {
+    try {
+      Config fallback = sysProps;
+      if (loadGlobalProperties && PathUtils.isAncestor(this.rootDirectory, path.getParent())) {
```

Comment:
Because the global properties of the current directory and all descendants are included automatically, however, we need to check whether we also need to include global properties of the parent directory and further ancestors.